src/main/java/org/rumbledb/compiler/TranslationVisitor.java

```diff
@@ -234,111 +232,110 @@ public Node visitExprSingle(JsoniqParser.ExprSingleContext ctx) {
                 returnExpr,
                 new ExceptionMetadata(
                         ctx.getStop().getLine(),
                         ctx.getStop().getCharPositionInLine()
                 )
         );
-        returnClause.setPreviousClause(previousFLWORClause);
+        previousFLWORClause.chainWith(returnClause);
 
         return new FlworExpression(
-                startClause,
-                contentClauses,
                 returnClause,
                 createMetadataFromContext(ctx)
         );
     }
 
     @Override
     public Node visitForClause(JsoniqParser.ForClauseContext ctx) {
-        List<ForClauseVar> vars = new ArrayList<>();
-        ForClauseVar child;
+        ForClause clause = null;
         for (JsoniqParser.ForVarContext var : ctx.vars) {
-            child = (ForClauseVar) this.visitForVar(var);
-            vars.add(child);
+            ForClause newClause = (ForClause) this.visitForVar(var);
+            if (clause != null) {
+                clause.chainWith(newClause.getFirstClause());
+            }
+            clause = (ForClause) newClause.getLastClause();
         }
 
-        return new ForClause(vars, createMetadataFromContext(ctx));
+        return clause;
     }
 
     @Override
     public Node visitForVar(JsoniqParser.ForVarContext ctx) {
         SequenceType seq = null;
         boolean emptyFlag;
-        VariableReferenceExpression var = (VariableReferenceExpression) this.visitVarRef(ctx.var_ref);
+        String var = ((VariableReferenceExpression) this.visitVarRef(ctx.var_ref)).getVariableName();
         if (ctx.seq != null) {
             seq = this.processSequenceType(ctx.seq);
         } else {
             seq = SequenceType.mostGeneralSequenceType;
         }
         emptyFlag = (ctx.flag != null);
-        VariableReferenceExpression atVarRef = null;
+        String atVar = null;
         if (ctx.at != null) {
-            atVarRef = (VariableReferenceExpression) this.visitVarRef(ctx.at);
+            atVar = ((VariableReferenceExpression) this.visitVarRef(ctx.at)).getVariableName();
         }
         Expression expr = (Expression) this.visitExprSingle(ctx.ex);
 
-        return new ForClauseVar(var, seq, emptyFlag, atVarRef, expr, createMetadataFromContext(ctx));
+        return new ForClause(var, emptyFlag, seq, atVar, expr, createMetadataFromContext(ctx));
     }
 
     @Override
     public Node visitLetClause(JsoniqParser.LetClauseContext ctx) {
-        List<LetClauseVar> vars = new ArrayList<>();
-        LetClauseVar child;
+        LetClause clause = null;
         for (JsoniqParser.LetVarContext var : ctx.vars) {
-            child = (LetClauseVar) this.visitLetVar(var);
-            vars.add(child);
+            LetClause newClause = (LetClause) this.visitLetVar(var);
+            if (clause != null) {
+                clause.chainWith(newClause.getFirstClause());
+            }
+            clause = (LetClause) newClause.getLastClause();
```

Comment:
Same comment as the forClause regarding method calls

Reply:
Very good catch! I simplified.